server/src/handlers/http/query.rs

```diff
@@ -64,81 +67,142 @@ pub struct Query {
     pub filter_tags: Option<Vec<String>>,
 }
 
+pub type QueryMap = Arc<Mutex<HashMap<u64, QueryStatus>>>;
 enum QueryStatus {
     Processing,
-    Result(Query),
+    Result(QueryResponse),
 }
 
-pub async fn query(req: HttpRequest, query_request: Query) -> Result<impl Responder, QueryError> {
+pub async fn query(
+    req: HttpRequest,
+    query_request: Query,
+    query_map: web::Data<QueryMap>,
+) -> Result<HttpResponse, QueryError> {
     let session_state = QUERY_SESSION.state();
 
-    // get the logical plan and extract the table name
+    // Generate hash for the query based on start, end, and query string
+    let hash = generate_hash(
+        &query_request.start_time,
+        &query_request.end_time,
+        &query_request.query,
+    );
+
+    // Check if the query is already being processed or completed
+    {
+        let query_map = query_map.lock().await;
+        if let Some(status) = query_map.get(&hash) {
+            match status {
+                QueryStatus::Processing => {
+                    // Wait for 55 seconds and check again
+                    sleep(Duration::from_secs(55)).await;
```

Comment:
this should be a env variable `P_PROXY_TIMEOUT` from user. Default to `60s`. Here we do `P_PROXY_TIMEOUT-5s`.

Reply:
okay, will change.